### ISSUE TITLE
Alternate FBX importer

### ISSUE THREAD

**Issue by ko-tengu** [Yellow-Dog-Man/Resonite-Issues#1498]
### Is your feature request related to a problem? Please describe.

The current FBX imports from the ASSIMP library are malformed and broken, and the library updates infrequently and slowly. If reverting the entire library to an older known working FBX version is undesirable or impossible, it may be better to have a dedicated FBX importer to work around this issue.

### Describe the solution you'd like

There are other lightweight FBX importer libraries, such as [OpenFBX](https://github.com/nem0/OpenFBX) or [FBXImport](https://github.com/WinteryFox/FBXImport) that may be able to be integrated into Resonite and used for this one file format.

### Describe alternatives you've considered

Converting FBX files to glTF or waiting for ASSIMP to update and fix the issues

### Additional Context

Related issues 
https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/1462
https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/1106
https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/813
https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/168

**Comment by Frooxius** [comment]
How do you know switching to these libraries would actually solve the problems? Integrating a whole new library is significant engineering task and unless you've done some testing with those particular problems, not only it's not guaranteed to solve these problems, but also introduce a lot of new ones.

You state that Assimp updates infrequently and slowly, but their last commit was 4 days ago, with a bunch more 2 weeks ago, 3 weeks ago, last month and so on. It's also one of the biggest, with 10K stars and 2.8K forks and 12409 total commits.

FBXImport in the meanwhile had last update 1 year ago and has only 16 stars. It doesn't even support all things that Assimp does, it still states that it has "Todo Animation Support". This makes me significantly doubt the maturity of this library.

OpenFBX seems a bit better, with last commit 2 weeks ago, but it still seems much smaller - just 186 commits, 1.1K stars, 124 forks.

Neither of the libraries seem like they would be better than Assimp or just fixing issues in Assimp itself - especially since the issues you mention are relatively minor ones and not major problems.

Unless you have more information on why either of those libraries would be better fix and would resolve the issues you provided, while also not risking introducing new major issues, we'll have to deny this request.

**Comment by ko-tengu** [comment]
This is still an issue with a lot of kinds of FBX files but in the time more tools have come out with common use by other open source game engines. 

ufbx is a small open source tool to convert ascii and binary FBX files, along with animations and .obj support that is very close to parity with the SDK.
https://github.com/ufbx/ufbx

FBX2GLTF is a layer that is built on the Autodesk FBX SDK and has full functionality but requires agreeing to the terms of the SDK.
https://github.com/godotengine/FBX2glTF

**Comment by nalathethird** [comment]
tbh, we are on a 7 year old version of Assimp.NET.

Assimp has updated MILES beyond whats available in terms of importing meshes and im pretty sure massive API changes and improvements overall. yes, i know the general rule with things like this is "dont fix it unless its broken", but honestly, Assimp needs an update. were currently on 5.0.0r1 from __May 2019__. *bruh*

__Assimp has been updated officially 30 TIMES since that release.__

I understand that we need bindings for Assimp -> Assimp.NET and thats what the repo is for, but honestly, the current Assimp.NET is using a crazy old version of .NET as well. sure, its fine for what we need, but, FrooxEngine could import better in terms of having that performance change from .NET 4.5 and .NET 2 to .NET 9 or even 10.

A different importer for FBX, would essentially mean that we need to change architecture entirely (like Froox said) its not a easy task and way out of scope for Resonite. 
Assimp is free, open source, and licensed properly for Resonite's needs, changing it now to something proprietary or something dedicated, would be something the team cant handle in their already overflowing issues in general. 

if anything comes from this issue, it should be that giving Assimp an update would be a good maintenance refresh and provide better support for models overall, possibly even fixing this in general to better support FBX (while its a format from 1990's, i dont see why we need "more support") But I notice that when i bring in massive scenes i notice it sometimes takes a few minuets (~1-10) and thats just odd to see from an importer for .NET (FrooxEngine does great with the data itself in terms of sending mesh, material and texture data to the renderer, its just the import process that takes a while)

**Comment by Frooxius** [comment]
@nalathethird That's actually not quite accurate:

- Assimp.NET is not Assimp itself - it's only the bindings for the actual library, meaning it only needs to change when the library ABI (its binary interface) changes - and that doesn't change that often
- It also doesn't significantly affect performance or import logic - it's literally just bindings of the library, not any of the actual import logic - that's within the native library
- We have our own fork of it that we keep updating whenever needed. Latest release is December 13, 2025: https://github.com/Yellow-Dog-Man/AssimpNet it's not as old as you think
- What matters is the native assimp library itself. The fork we're using is currently about 2 years old - we had to revert to an older one due to newer versions breaking FBX blendshape order: https://github.com/Yellow-Dog-Man/assimp
- Sure it's 2 years old, but that's not 7 years when it matters for the import

**Comment by Banane9** [comment]
> I understand that we need bindings for Assimp -> Assimp.NET and thats what the repo is for, but honestly, the current Assimp.NET is using a crazy old version of .NET as well. sure, its fine for what we need, but, FrooxEngine could import better in terms of having that performance change from .NET 4.5 and .NET 2 to .NET 9 or even 10.

And to add on to what Froox said - that's also not how .NET works. The version something targets doesn't really matter for performance, only the version of the runtime it's, well, run in.

**Comment by Frooxius** [comment]
Also to respond to @ko-tengu:

I understand that Assimp has a lot of issues with FBX. I'm not contesting that.

However you have to realize that switching to a different importer system is a substantial amount of work and likely bring a number of its own teething issues that we'll have to deal with.

So if we were to commit that time into it - I would appreciate some actual data/information on reliability of those other solutions.

Otherwise you're asking us to commit a good chunk of our time into something that might not solve the problem at all or perhaps have a whole different set of problems and we'd have to revert. We got burned on this quite a few times in the past, which is why we want to have some more confidence on a solution before we pull trigger on it. 

The "current solution has a lot of problems" is not enough for us to commit, because we don't know if the new solution will also have a lot of problems - we'd like to know a bit more about it and some experience with other projects first.

All I'm asking is to be respectful of our time - we have very limited amount of it and we need to be smart about where we invest it. Diving head-first into another solution without knowing that it'll likely fix our issues is would not be smart on our end.

---------

TL;DR: If you want us to move on this, do these:

- Give us some data/examples on import reliability of alternatives
- Do any big projects use any of the alternatives?
- Do you have better success rate with variety of models with any of the alternatives?
- Are the alternatives actively developed? Are there signs that they won't be abandoned?
- Generally anything that will help build confidence in the alternate solution

Don't do this:
- Re-state that Assimp has issues 
     - For all we know the alternatives can have even more issues right now
     - This won't convince us to commit time

**Comment by lxw404** [comment]
Just to mention: after a full year of deliberations, the blendshape re-ordering fix has finally been pulled into assimp's current main branch which also has accumulated a multitude of other fixes for the FBX format:

https://github.com/assimp/assimp/commit/392a658f9c271be965271f45e7521a1b80ea4392

It might be worth smoke testing with some standard model files to make sure it's okay to update to.

**Comment by ko-tengu** [comment]
uFBX is extremely battletested at this point, being the default FBX importer for Godot for the [last two years.](https://github.com/godotengine/godot/pull/81746) It does import animations and blendshapes correctly, along with the older FBX formats. With personal testing, models and animations that give Resonite trouble do import fine into newer Godot versions that use uFBX (4.3+.)

The main missing feature uFBX has compared to Assimp would be support for exporting FBX files, though this is under active development.
